app/services/ui_notifications.rb

```diff
@@ -0,0 +1,37 @@
+# Namespace for Foreman UI notifications event handling
+module UINotifications
+  class Base
+    attr_reader :subject, :event
+
+    def self.deliver!(subject, event)
+      new(subject, event).deliver!
+    end
+
+    def initialize(subject, event)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
+      raise(Foreman::Exception, 'invalid event') unless allowed_events.include?(event)
+      @subject = subject
+      @event = event
+    end
+
+    def deliver!
+      logger.debug("Notification event: #{event} - checking for notifications")
+      send(event) if respond_to?(event, true)
+    end
+
+    protected
+
+    # Defaults to anonymous api admin, override in subclasses as needed.
+    def initiator
+      User.anonymous_api_admin
+    end
+
+    def allowed_events
```

Comment:
I would like to see a DSL for extending this list instead of overriding the method. Good example would be `Host` events that could be extended by Katello to add some content specific events.